### PR TITLE
component: add initial Google post processing component and topology

### PR DESCRIPTION
**Pull request by kehuangli**
Introduce a new component to perform configurable audio post processing
on playback path before smart amplifier.

1. Put google_audio_post_processing libs at $SOF_REPO/third_party/lib
   - libgoogle_audio_post_processing.a
   - libc++.a
   - libc++abi.a
2. Enable google_audio_post_processing in config override
   $SOF_REPO/src/arch/xtensa/configs/override
   CONFIG_COMP_GOOGLE_AUDIO_POST_PROCESSING=y
3. Build the firmware and tool with xcc
4. To verify it works:
   - aplay some audio
   - amixer and grep GAPP
   - sof-ctl to get/set control

Signed-off-by: Kehuang Li <kehuangli@google.com>

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by marc-hb** [comment]
test this please

File: src/audio/google_audio_post_processing.c
Comment (dbaluta) [inline src/audio/google_audio_post_processing.c:3]: I think it will be worth thinking on creating a separate vendor directory in audio for each specific component. 

E.g ```/src/audio/google/```

File: src/audio/google_audio_post_processing_mock.c
Comment (dbaluta) [inline src/audio/google_audio_post_processing_mock.c:2]: I'm pretty sure this is not the correct path to place the mock files. Maybe: ./test/ directory can be used for that?

File: src/audio/google_audio_post_processing_mock.c
Comment (dbaluta) [inline src/audio/google_audio_post_processing_mock.c:29]: preffered way would be call `rballoc` later after `s` declaration. ANd also check for return value.

File: src/audio/google_audio_post_processing.c
Comment (cujomalainey) [reply src/audio/google_audio_post_processing.c:3]: Agreed, I will do that as a separate PR after this lands

File: src/audio/google_audio_post_processing_mock.c
Comment (cujomalainey) [reply src/audio/google_audio_post_processing_mock.c:2]: I disagree, that is for running tests, these are not tests but stand ins that can be used during development

File: src/audio/google_audio_post_processing_mock.c
Comment (cujomalainey) [reply src/audio/google_audio_post_processing_mock.c:29]: What is wrong with instantiating and declaring in the same line? But yes we do need a nullity check here

**Review by cujomalainey — CHANGES_REQUESTED** [review]

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:322]: looks like init is not required

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:262]: dev can be moved to the prev line

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:260]: better to replace assert with an error check and return -ENODEV instead

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:341]: no check for memcpy_s failure here?

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:392]: might be simpler to just return comp_verify_params() directly. An error is printed in comp_verify_params() already.

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:403]: no init needed

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:419]: one line?

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:422]: same here

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:426]: here too

File: src/audio/google_audio_post_processing_mock.c
Comment (ranj063) [inline src/audio/google_audio_post_processing_mock.c:48]: why not return the return value or memcpy_s?

File: src/audio/google_audio_post_processing_mock.c
Comment (ranj063) [inline src/audio/google_audio_post_processing_mock.c:61]: better to check for return value

File: src/audio/google_audio_post_processing_mock.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing_mock.c:29]: Added a check.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c:327]: I am not sure about this comment, do you mean the whole prepare() function is not required (cuz there is no init in the highlighted lines), but in prepare() we check values and do setup/configuration, do you mean all should be moved to new()?

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Updated the printed info, and it's now moved to prev line.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: replaced, thx.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Added check for all memcpy_s. I did miss some, thx for catching the issue.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: makes sense, updated.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c:418]: Don't we have some line length restriction? if making this one line, it will be a long line > 80 characters.

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [reply src/audio/google_audio_post_processing.c:327]: no I meant ret doesnt need to be initialized to 0

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [reply src/audio/google_audio_post_processing.c:418]: 100 Chars is the limit

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: I see. fixed.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Switched to 100 in clang-format. It changed the whole file a lot due to the new length limit.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing_mock.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing_mock.c]: Done.

File: src/audio/google_audio_post_processing_mock.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing_mock.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (ranj063) [inline src/audio/google_audio_post_processing.c:306]: this check for ret probably needs to be within the if {} above

**Review by ranj063 — COMMENTED** [review]
looks good @kehuangli. just a minor comment from me now

**Review by ranj063 — APPROVED** [review]

File: src/audio/google_audio_post_processing_mock.c
Comment (dbaluta) [reply src/audio/google_audio_post_processing_mock.c:29]: @cujomalainey nothing wrong but the code is harder to read. Anyhow, not a blocker anway.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Good catch, fixed.

**Comment by lgirdwood** [comment]
@cujomalainey I assume you will add @kehuangli to sof-developers and google GH groups ?

**Comment by cujomalainey** [comment]
@lgirdwood thanks for the reminder

File: src/audio/google_audio_post_processing.c
Comment (bzhg) [inline src/audio/google_audio_post_processing.c:280]: comp_new signature and the way ipc config being passed in have changed since tgl-013.
#include <sof/audio/ipc-config.h>
...
static struct comp_dev *gapp_new(const struct comp_driver *drv, struct comp_ipc_config *config,void *spec) {
struct ipc_config_process *ipc_gapp = spec;
dev = comp_alloc(drv, sizeof(*dev));
dev->ipc_config = *config;
...
}

File: src/audio/google_audio_post_processing.c
Comment (bzhg) [inline src/audio/google_audio_post_processing.c:461]: buffer_stream_invalidate

File: src/audio/google_audio_post_processing.c
Comment (bzhg) [inline src/audio/google_audio_post_processing.c:465]: buffer_stream_writeback

File: src/audio/google_audio_post_processing.c
Comment (bzhg) [inline src/audio/google_audio_post_processing.c:309]: Blob related functions have been moved to:
#include <sof/audio/data_blob.h>

**Review by bzhg — APPROVED** [review]
A few updates needed to run on sof/main since there has been some api changes from tgl-013 to main. I can submit a follow up patch to update these.

**Comment by kehuangli** [comment]
@bzhg That will be better if fix in a separate PR, since we could CP this patch back to tgl-013 w/o conflict.

**Comment by dbaluta** [comment]
@kehuangli yes, but you won't merge this PR as it is as this will break compilation.

**Review by lgirdwood — APPROVED** [review]
@cujomalainey pls feel free to merge once you are happy.

File: src/audio/google_audio_post_processing.c
Comment (lyakh) [inline src/audio/google_audio_post_processing.c:108]: is a log entry really needed every time volume is requested?

File: src/audio/google_audio_post_processing.c
Comment (lyakh) [inline src/audio/google_audio_post_processing.c:76]: wouldn't `comp_dbg()` be enough? Particularly since you only log the fact of this command being sent, not even any volume values?

File: src/audio/google_audio_post_processing.c
Comment (lyakh) [inline src/audio/google_audio_post_processing.c:240]: you need a semicolon after `COMPILER_FALLTHROUGH`, but in fact I don't think you need it here at all: there are many empty equivalent `case` actions like
```
switch (x) {
case A:
case B:
case C:
case D:
case E:
	do_something;
}
```
and you don't want to put `COMPILER_FALLTHROUGH` in each of them. IMHO only if there are some statements in such a `case` entry, then you need a `COMPILER_FALLTHROUGH`.

File: src/audio/google_audio_post_processing_mock.c
Comment (lyakh) [inline src/audio/google_audio_post_processing_mock.c:100]: looks like you actually want to check for `num_channels > ARRAY_SIZE(state->volume)`

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Replaced all comp_info to comp_dbg.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Removed.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Good point. Updated.

File: src/audio/google_audio_post_processing.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing.c]: Done.

File: src/audio/google_audio_post_processing_mock.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing_mock.c]: Good catch. easy to make mistake with sizeof.

**Comment by kehuangli** [comment]
@[dbaluta](https://github.com/dbaluta) All the code in this PR is guarded by config flag, so it won't break main build, and I have built and tested on tgl-013 branch with flag on. A follow up PR from bzhg will allow main branch to turn on the flag, when this current one won't have observable effect to main at this point.

**Review by lyakh — APPROVED** [review]
Looks better now, thanks!

**Comment by lgirdwood** [comment]
@cujomalainey @bzhg good for you guys ?

**Comment by cujomalainey** [comment]
@lgirdwood both @bzhg and I are OOO, please merge when you are content with it, all my issues have been resolved in internal reviews.

**Comment by lgirdwood** [comment]
@singalsu any comments ?

File: src/audio/google_audio_post_processing_mock.c
Comment (singalsu) [inline src/audio/google_audio_post_processing_mock.c:17]: This is second time define for the same, move to google_audio_post_processing.h?

**Review by singalsu — APPROVED** [review]

File: src/audio/google_audio_post_processing_mock.c
Comment (kehuangli) [reply src/audio/google_audio_post_processing_mock.c:17]: Actually, I did that on purpose. 48kHz is a general default, but it may not always be the case, so I'd prefer not to define it in the library API. Therefore, later if encountering special cases, e.g., 96kHz or 16kHz, we won't need to change the const in the API; instead, we can update the API to pass-in the sample rate and increase the API version number. WDYT?

Comment (singalsu) [reply src/audio/google_audio_post_processing_mock.c:17]: Yep that makes sense.

**Pull request merged**
